### PR TITLE
New event handler

### PR DESCRIPTION
**Pull request by edwardtfn**
This rebuild the engine to send events to Home Assistant

## Summary by CodeRabbit

- **New Features**
  - Introduced a new API section for enhanced event handling.
  - Added scripts for sending Home Assistant events on device boot.
  - Implemented new binary sensor configurations for touch events, including click, double-click, and long-press actions.
  - Enhanced button handling logic for improved action clarity and functionality.
  - Added new scripts for initializing touch formats and gangs.

- **Bug Fixes**
  - Improved error logging for relay control and button interactions.

- **Documentation**
  - Enhanced README for better clarity and structure.
  - Added comprehensive documentation for the event system in the firmware.

- **Chores**
  - Removed outdated API configuration file.

### CODE REVIEW

File: ESPHome/TX-Ultimate-Easy-ESPHome_core_hw_buttons.yaml [hunk 1]
```diff
@@ -108,33 +108,89 @@ script:
...
+              case ${BUTTON_4_ID}:
+                bs_button_4_long_press_event->publish_state(true);
+                break;
+            }
+          }
           id(button_press_button) = 0;
-          id(click_counter) = 0;
+          id(button_press_position) = 0;
           id(button_press_start_time) = 0;
+          id(click_counter) = 0;
```
[the remark this answers is not in the document]
Comment (edwardtfn) [reply ESPHome/TX-Ultimate-Easy-ESPHome_core_hw_buttons.yaml:177]: I will look at this in the future if this becomes an issue.
As this is about manual actions in a small device, the chances of race issues are very low.

File: ESPHome/TX-Ultimate-Easy-ESPHome_core_hw_touch.yaml [hunk 2]
```diff
@@ -15,50 +15,188 @@
...
+    name: Button 4 - Long-press event
+    internal: false
+    <<: *bs_button_event_template
+    on_press:
+      then:
+        - delay: !lambda return nr_touch_duration->state;
+        - binary_sensor.template.publish:
+            id: bs_button_4_long_press_event
+            state: OFF  # yamllint disable-line rule:truthy
+
```
[the remark this answers is not in the document]
Comment (edwardtfn) [reply ESPHome/TX-Ultimate-Easy-ESPHome_core_hw_touch.yaml:147]: The proposal won't work with ESPHome. I will keep it as it is for now.

File: ESPHome/TX-Ultimate-Easy-ESPHome_core_hw_buttons.yaml [hunk 3]
```diff
@@ -157,30 +213,24 @@ script:
...
-                  if (button > model_idx)
-                    button = model_idx; // Clamp to max button count
-              }
+          id(button_press_button) = button;
+          // Update counters
+          if (id(button_press_position) == position) {
+            id(click_counter)++;
+          } else {
+            id(click_counter) = 1;
+            id(button_press_position) = position;
```
[the remark this answers is not in the document]
Comment (edwardtfn) [reply ESPHome/TX-Ultimate-Easy-ESPHome_core_hw_buttons.yaml:233]: Button id validation exists on the component side.

File: ESPHome/TX-Ultimate-Easy-ESPHome_core.yaml [hunk 4]
```diff
@@ -144,6 +160,9 @@ select:
     entity_category: config
     disabled_by_default: false
     icon: mdi:dip-switch
+    on_value:
+      then:
+        - lambda: set_gang_count(static_cast<uint8_t>(i)+1);
```
[the remark this answers is not in the document]
Comment (edwardtfn) [reply ESPHome/TX-Ultimate-Easy-ESPHome_core.yaml]: This validation exists on the component side.

**Comment by cavazosomar** [comment]
Seems like with the recent changes the time it takes from a button release to a switch toggle state has increased noticeably. Even the logs show that operations are taking too long, perhaps some tuning is required.
Screenshot_20241225_210812.

Also, can it be possible to implement 'on press' logic for the buttons? I know gestures wouldn't work but I imagine a lot of logic could be skipped this way increasing response time greatly

**Comment by edwardtfn** [comment]
The delay is due to the logic for waiting for a second (and consecutive) clicks to be able to detect a `double-click` event.

I've later added a substitution for that, so probably reducing `BUTTON_MULTI_CLICK_DELAY` will make it harder to detect a double click but will make it more responsive to single clicks:
https://github.com/edwardtfn/TX-Ultimate-Easy/blob/4378952d979662138dcb8f8844a5d69a471fbf76/ESPHome/TX-Ultimate-Easy-ESPHome_core_hw_buttons.yaml#L27-L29

If you don't mind playing a bit with those substitutions and share here the values that fits your better, that would be awesome. 😉

**Comment by cavazosomar** [comment]
Should these substitutions work from my main yaml file? I suspect not because I added those lines and modified the values but the default values are still in effect

substitutions:
  name: <my-device-name>
  friendly_name: <"my friendly name">
  api_key: !secret api_key
  ota_password: !secret ota_password
  BUTTON_CLICK_MIN_LENGTH: '30'    # The minimum duration the click should last, in msec
  BUTTON_CLICK_MAX_LENGTH: '250'   # The maximum duration the click should last, in msec
  BUTTON_MULTI_CLICK_DELAY: '150'  # The time to wait for another click, in msec
  BUTTON_PRESS_TIMEOUT: '10000'    # Ignore if button is pressed for longer than this time, in msec
  BUTTON_LONG_PRESS_DELAY: '800'   # The time to wait to consider a long press, in msec

...

**Comment by edwardtfn** [comment]
Yes, it should work from your file.
Do you mind sharing your device's yaml? Please remove any sensitive info before sharing.

**Comment by cavazosomar** [comment]
This is my yaml config file:
tx-ultimate-3g.yaml

---
substitutions:
  name: <my-device-name>
  friendly_name: <"my friendly name">
  api_key: !secret api_key
  ota_password: !secret ota_password
  BUTTON_CLICK_MIN_LENGTH: '30'    # The minimum duration the click should last, in msec
  BUTTON_CLICK_MAX_LENGTH: '250'   # The maximum duration the click should last, in msec
  BUTTON_MULTI_CLICK_DELAY: '150'  # The time to wait for another click, in msec
  BUTTON_PRESS_TIMEOUT: '10000'    # Ignore if button is pressed for longer than this time, in msec
  BUTTON_LONG_PRESS_DELAY: '800'   # The time to wait to consider a long press, in msec

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

#append encryption
api:
  encryption:
    key: "${api_key}"

#append password 
ota: 
#- platform: esphome
  password: "${ota_password}"

  
packages:
  remote_package:
    url: https://github.com/edwardtfn/TX-Ultimate-Easy
    ref: main
    refresh: 30s
    files:
      - ESPHome/TX-Ultimate-Easy-ESPHome_core.yaml
      # - ESPHome/TX-Ultimate-Easy-ESPHome_addon_ble_proxy.yaml  # Adds experimental BLE proxy support (limited)

**Comment by edwardtfn** [comment]
Please try this:
```yaml
substitutions:
  name: <my-device-name>
  friendly_name: <"my friendly name">
  api_key: !secret api_key
  ota_password: !secret ota_password
  BUTTON_CLICK_MIN_LENGTH: '30'    # The minimum duration the click should last, in msec
  BUTTON_CLICK_MAX_LENGTH: '250'   # The maximum duration the click should last, in msec
  BUTTON_MULTI_CLICK_DELAY: '150'  # The time to wait for another click, in msec
  BUTTON_PRESS_TIMEOUT: '10000'    # Ignore if button is pressed for longer than this time, in msec
  BUTTON_LONG_PRESS_DELAY: '800'   # The time to wait to consider a long press, in msec

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

#append encryption
api:
  encryption:
    key: "${api_key}"

#append password 
ota: 
#- platform: esphome
  password: "${ota_password}"

  
packages:
  remote_package:
    url: https://github.com/edwardtfn/TX-Ultimate-Easy
    ref: main
    refresh: 30s
    files:
      - ESPHome/TX-Ultimate-Easy-ESPHome.yaml
      # - ESPHome/TX-Ultimate-Easy-ESPHome_addon_ble_proxy.yaml  # Adds experimental BLE proxy support (limited)
```

**Comment by edwardtfn** [comment]
Hi,
Any success with changes to `BUTTON_MULTI_CLICK_DELAY`?

**Pull request merged**